docs/src/tutorial/minimal_working_compute.jl

```diff
@@ -0,0 +1,266 @@
+                            "main", # this needs to match the function name in the shader
+                            specialization_info = SpecializationInfo(
+                                [SpecializationMapEntry(0, 0, 4)],
+                                UInt64(4),
+                                Ptr{Nothing}(pointer(spec_consts)),
+                            ),
+                        ),
+                        pl,
+                        -1,
+                    ),
+                ],
+            ),
+        ),
+    ),
+)
+
+# Now make a descriptor pool that is used to allocate the buffer descriptors
+# from (not a big one, just 1 descriptor set and total 1 descriptor)
+dpool = unwrap(create_descriptor_pool(device, 1, [DescriptorPoolSize(DESCRIPTOR_TYPE_STORAGE_BUFFER, 1)]))
+
+# Allocate the descriptors for our layout
+dsets = unwrap(allocate_descriptor_sets(device, DescriptorSetAllocateInfo(dpool, [dsl])))
+dset = first(dsets)
+
+# Make the descriptors point to the right buffers
+update_descriptor_sets(
+    device,
+    [
+        WriteDescriptorSet(
+            dset,
+            0,
+            0,
+            DESCRIPTOR_TYPE_STORAGE_BUFFER,
+            [],
+            [DescriptorBufferInfo(buffer, 0, WHOLE_SIZE)],
+            [],
+        ),
+    ],
+    [],
+)
+
+# ## Executing the shader
+#
+# Let's create a command pool in the right queue family, and take a command
+# buffer out of that.
+cmdpool = unwrap(create_command_pool(device, qfam_idx))
+cbufs = unwrap(allocate_command_buffers(device, CommandBufferAllocateInfo(cmdpool, COMMAND_BUFFER_LEVEL_PRIMARY, 1)))
+cbuf = first(cbufs)
+
+# Now that we have a command buffer, we can fill it with commands that cause
+# the kernel to be run. Basically, we bind and fill everything, and then
+# dispatch a sufficient amount of invocations of the shader to span over the
+# array.
+begin_command_buffer(cbuf, CommandBufferBeginInfo(flags = COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT))
+
+cmd_bind_pipeline(cbuf, PIPELINE_BIND_POINT_COMPUTE, p)
+
+const_buf = [shader_push_consts(1.234, data_items)]
+cmd_push_constants(cbuf, pl, SHADER_STAGE_COMPUTE_BIT, 0, sizeof(shader_push_consts), Ptr{Nothing}(pointer(const_buf)))
+
+cmd_bind_descriptor_sets(cbuf, PIPELINE_BIND_POINT_COMPUTE, pl, 0, [dset], [])
+
+cmd_dispatch(cbuf, div(data_items, const_local_size_x, RoundUp), 1, 1)
+
+end_command_buffer(cbuf)
+
+# Finanly, find a handle to the compute queue and send the command to execute the shader!
+compute_q = get_device_queue(device, qfam_idx, 0)
+queue_submit(compute_q, [SubmitInfo([], [], [cbuf], [])])
+
+# ## Getting the data
+#
+# Now, data is being crunched in the background. To get the data, we need to
+# wait for completion and invalidate the mapped memory (so that whatever
+# happened on the GPU gets to the mapped range visible for the "host" device.
+queue_wait_idle(compute_q)
+invalidate_mapped_memory_ranges(device, [MappedMemoryRange(mem, 0, mem_size)])
```

Comment:
```suggestion
# Wait for computations to be carried out, while making sure data is kept alive
# during queue operations (in non-global scopes, such as functions, the compiler
# may skip the allocation of unused variables or garbage-collect variables that the
# runtime thinks are unused. Notably, the runtime is not aware that there is a memory
# dependency with these variables until that command returns, so we tell it manually.
GC.@preserve buff dsl pl p dpool dset cmdpool cbuf const_buf begin
    queue_wait_idle(compute_q)
end

# Note that the invalidation is only required for memory that is not host-coherent,
# so you may not need this step if your memory has the right property flag.
invalidate_mapped_memory_ranges(device, [MappedMemoryRange(mem, 0, mem_size)])
```

Reply:
wow, how come this @preserve is needed?

Reply:
That's to prevent the GC from running, which in turn prevents finalizers from running, those same finalizers which can call `vkFreeMemory` or `vkDestroy...` functions. One downside of automatic memory management.

Reply:
Note that if you were to call `finalize(mem); finalize(buff); ...` etc yourself after all commands have ended, it would have the same effect, and finalizers would be run only once.

Reply:
why would it destroy say the `buff` if it stays actively in the scope? :thinking:

Reply:
(technically the same could happen in `queue_submit` and `allocate_memory` functions as they take some time too, right?)

Reply:
> why would it destroy say the buff if it stays actively in the scope?

In global scope, all variables are kept alive (unless overwritten), so running the tutorial in global scope is fine. But if you were to put everything into a `main()` function, and call the function, you'd get errors.

The thing is that from the point that `buff` is not reused further (so, anytime from the last statement that used `buff`), the GC can reclaim it.

Reply:
`queue_submit` and `allocate_memory` don't have issues because they preserve all required variables until the call returns (if you want some more info, you can check out [these docs](https://juliagpu.github.io/Vulkan.jl/dev/about/interfacing/), but I should probably add even more info to the Vulkan docs related normal use, like here)

Reply:
(making a GC at each sync looks like an overkill to me)

Reply:
oic that's basically what `@preserve` does. I'll get that updated.

Reply:
Don't worry, `GC.@preserve` doesn't trigger the GC, it's just that if the GC happens to run for whatever reason, all preserved variables are guaranteed to not be collected (by a process called "rooting", where we declare that these variables are unconditionally alive).

But to avoid interacting with GC explicitly, you can just `finalize` all the variables manually (instead of relying on the GC to do that for you). In this way, you 1. preserve all variables until the `finalize` calls are executed, and 2. make sure that you release the resources early (without waiting for the GC to trigger again in the future and only then run the finalizers).

Usually, in production environment, that's less of an issue since you're going to store handles/resources.